chainercv/links/model/fpn/mask_utils.py

```diff
@@ -0,0 +1,181 @@
+
+    # To work around an issue with cv2.resize (it seems to automatically
+    # pad with repeated border values), we manually zero-pad the masks by 1
+    # pixel prior to resizing back to the original image resolution.
+    # This prevents "top hat" artifacts. We therefore need to expand
+    # the reference boxes by an appropriate factor.
+    backend = resize_backend()
+    if backend == 'cv2':
+        padded_segm_size = segm_size + pad * 2
+        expand_scale = padded_segm_size / segm_size
+        bbox = _expand_bbox(bbox, expand_scale)
+        resize_size = padded_segm_size
+        slice_ = slice(pad, -pad)
+    else:
+        resize_size = segm_size
+        slice_ = slice(0, segm_size)
+    bbox = _integerize_bbox(bbox)
+
+    segm = []
+    if index is None:
+        index = np.arange(len(bbox))
+    else:
+        index = chainer.backends.cuda.to_cpu(index)
+
+    for i, bb in zip(index, bbox):
+        y_min = max(bb[0], 0)
+        x_min = max(bb[1], 0)
+        y_max = max(min(bb[2], H), 0)
+        x_max = max(min(bb[3], W), 0)
+        if y_max - y_min == 0 or x_max - x_min == 0:
+            segm.append(np.zeros((segm_size, segm_size), dtype=np.float32))
+            continue
+
+        bb_height = bb[2] - bb[0]
+        bb_width = bb[3] - bb[1]
+        cropped_m = np.zeros((bb_height, bb_width), dtype=np.bool)
+
+        y_offset = y_min - bb[0]
+        x_offset = x_min - bb[1]
+        cropped_m[y_offset:y_offset + y_max - y_min,
+                  x_offset:x_offset + x_max - x_min] =\
+            chainer.backends.cuda.to_cpu(mask[i, y_min:y_max, x_min:x_max])
+        sgm = transforms.resize(
+            cropped_m[None].astype(np.float32),
+            (resize_size, resize_size))[0].astype(np.int32)
+        segm.append(sgm[slice_, slice_])
+
+    return np.array(segm, dtype=np.float32)
+
+
+def segm_to_mask(segm, bbox, size):
+    """Recover mask from cropped and resized mask.
+
+    Args:
+        segm (~numpy.ndarray): See below.
+        bbox (~numpy.ndarray): See below.
+        size (tuple): This is a tuple of length 2. Its elements are
+            ordered as (height, width).
+
+    Returns:
+        ~numpy.ndarray: See below.
+
+    .. csv-table::
+        :header: name, shape, dtype, format
+
+        :obj:`segm`, ":math:`(R, S, S)`", :obj:`float32`, --
+        :obj:`bbox`, ":math:`(R, 4)`", :obj:`float32`, \
+        ":math:`(y_{min}, x_{min}, y_{max}, x_{max})`"
+        :obj:`mask` (output), ":math:`(R, H, W)`", :obj:`bool`, --
+
+    """
+    pad = 1
+    H, W = size
+    _, segm_size, _ = segm.shape
+
+    mask = np.zeros((len(bbox), H, W), dtype=np.bool)
+
+    # To work around an issue with cv2.resize (it seems to automatically
+    # pad with repeated border values), we manually zero-pad the masks by 1
+    # pixel prior to resizing back to the original image resolution.
+    # This prevents "top hat" artifacts. We therefore need to expand
+    # the reference boxes by an appropriate factor.
```

Comment:
We have two similar blocks. Can we merge them?

Reply:
OK